### PR TITLE
Refactor EvaluatorTestBase To Allow Testing of Multiple Compilers

### PR DESCRIPTION
**Pull request by dlurton**
The purpose of this PR is to prepare the unit test suite to be able to test multiple compilers.

I have confidence in this changes because the code coverage report differs by only a fraction of a percent with these changes.

You'll want to check this out locally and start with the `EvaluatorTestBase` (ETB) class which was entirely rewritten, then look at `EvaluatorTestAdapter` and implementations of it.  The other changes are (more or less) mechanical refactors that call into the new ETB functions (the bulk of the diffs in this PR).  **However, I avoided changing the test cases themselves like the plague.**

For ETB, there used to be about 8 protected functions (including overloads) that were for:

- "happy path" scenarios (i.e. given a query and a session, assert a particular value was returned), or...
- error scenarios (i.e. given the same, assert a particular error was thrown)

Each of these functions were slightly different.  I replaced all of these functions in ETB with `runEvalutorTestCase` and `runEvaluatorErrorTestCase` which have all the capabilities of the prior functions and updated all of the old call sites to call one of these two functions.  One of the challenges here is that, for the "happy path" tests, there are 4 different formats for expected values (see: `ExpectedResultFormat` to see how I handled this... as mentioned above, I avoided changing the test cases themselves... otherwise, this would have been far too big of a task for me to handle alone).

Each of the new ETB functions creates an instance of `EvaluatorTestCase` or `EvaluatorErrorTestCase` which contain the data for the test case and passes it to the appropriate function on `EvaluatorTestAdapter` to perform the actual tests and assertions.

`EvaluatorTestAdapter` is an interface with a few implementations--one runs the test in multiple other adapters, one that runs the test using the AST evaluator, one that tests the legacy serializers, etc...  In the near future, an implementation of `EvaluatorTestAdapter` will test the new query planner & physical plan compiler.

Also, for the test DSL ([example use](https://github.com/partiql/partiql-lang-kotlin/blob/51e7da7b5e63e45f01c4df101168b2117a17a2d1/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L106)) was also are modified to use a `EvaluatorTestAdapter`.

## One Idea for Future Use of  `EvaluatorTestAdapter`

In theory, an implementation of `EvaluatorTestAdapter` could be created that emits test case data as Ion, possibly as input to a "compliance suite" of some sort.  This would be a first step toward converting these unit tests to "test scripts" that could then be used to test other implementations of PartiQL.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/eval/Thunk.kt [hunk 1]
```diff
@@ -526,24 +549,7 @@ internal class LegacyThunkFactory(
...
-                else -> {
-                    // Only add source location data to the error context if it doesn't already exist
-                    // in [errorContext].
-                    if (!e.errorContext.hasProperty(Property.LINE_NUMBER)) {
-                        sourceLocation?.let { fillErrorContext(e.errorContext, sourceLocation) }
-                    }
-                    throw e
-                }
-            }
+            throw populateErrorContext(e, sourceLocation)
```
Comment (dlurton) [inline lang/src/org/partiql/lang/eval/Thunk.kt:552]: Corrects a small bug I discovered where line/column information was not being added to some `EvaluationException`s.

File: lang/test/org/partiql/lang/TestBase.kt [hunk 2]
```diff
@@ -16,213 +16,15 @@
...
-        }
-
-        assertThat(propertyValue)
-            .withFailMessage(message)
-            .isEqualTo(entry.value)
-    }
-}
 
 @RunWith(JUnitParamsRunner::class)
 abstract class TestBase : Assert() {
```
Comment (dlurton) [inline lang/test/org/partiql/lang/TestBase.kt:26]: The old `TestBase` class contained a bunch of protected functions that weren't being used except in a small handful of places anymore so I decided to remove them and refactor the uses.

File: lang/test/org/partiql/lang/eval/BindingsTest.kt [hunk 3]
```diff
@@ -14,29 +14,29 @@
...
-        ).run(block)
+        val delegatedBindings = innerBindings.delegate(outerBindings)
 
-    @Test
-    fun delegate() = over("{a:1, b:2}", { this.delegate(bind("{b:3, c:4}")) }) {
-        assertBinding("b") { ion.newInt(2) == ionValue }
-        assertBinding("c") { ion.newInt(4) == ionValue }
+        // note that "b" in the inner bindings is shadowed by the "b" in the outer bindings.
+        assertEquals(2, delegatedBindings[BindingName("b", BindingCase.INSENSITIVE)]!!.intValue())
+        assertEquals(4, delegatedBindings[BindingName("c", BindingCase.INSENSITIVE)]!!.intValue())
```
Comment (dlurton) [inline lang/test/org/partiql/lang/eval/BindingsTest.kt:39]: This was used old `TestBase` stuff--the refactor tests the same thing but without the dependency on test base.

File: lang/test/org/partiql/lang/TestBase.kt (hunk 2 above)
Comment (dlurton) [reply lang/test/org/partiql/lang/TestBase.kt:26]: It's also worth noting that `TestBase` still inherits from `org.junit.Assert`, which is how many of the tests access the `assert*` methods (through `Assert`'s `protected` methods.)  I really dislike this since these functions shadow anything `import`ed, which creates some confusing situations sometimes in the IDE's code completion.  This also means we have many JUnit5 tests that are using the JUnit4 assertion functions.  This is not a huge problem, just a bit messy.  I considered fixing that to be out of scope for this work, which is why `TestBase` wasn't removed.

File: lang/test/org/partiql/lang/eval/CastTestBase.kt [hunk 4]
```diff
@@ -56,6 +56,46 @@ data class FixSemantics(val expectedQuality: CastQuality) : CastQualityStatus()
 
 abstract class CastTestBase : EvaluatorTestBase() {
 
+    fun ConfiguredCastCase.assertCase(
```
Comment (dlurton) [inline lang/test/org/partiql/lang/eval/CastTestBase.kt:59]: Needed to move this here so we could access the `runEvaluator[Error]TestCase` functions of the new `EvaluatorTestBase`.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt [hunk 5]
```diff
@@ -20,30 +20,12 @@ class EvaluatingCompilerDateTimeTests : EvaluatorTestBase() {
...
-        private fun case(query: String, expected: String) = EvaluatorTestCase(query, expected)
-
-        override fun getParameters() = listOf(
-            case("DATE '2012-02-29'", "2012-02-29"),
-            case("DATE '2021-02-28'", "2021-02-28"),
-            case("DATE '2021-03-17' IS DATE", "true"),
-            case("'2021-03-17' IS DATE", "false")
+    @Test
+    fun testDateLiteral() {
+        runEvaluatorTestCase(
```
Comment (dlurton) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt:25]: This test didn't really need to be parameterized and had redundant assertions.

File: lang/test/org/partiql/lang/syntax/SqlParserTestBase.kt [hunk 6]
```diff
@@ -43,6 +49,20 @@ abstract class SqlParserTestBase : TestBase() {
 
     protected fun parse(source: String): PartiqlAst.Statement = parser.parseAstStatement(source)
 
+    private fun assertSexpEquals(
```
Comment (dlurton) [inline lang/test/org/partiql/lang/syntax/SqlParserTestBase.kt:52]: This comes from `TestBase`... since this class was the only place it was being used, I moved it here.

File: lang/test/org/partiql/lang/syntax/SqlParserTestBase.kt [hunk 7]
```diff
@@ -207,3 +227,88 @@ abstract class SqlParserTestBase : TestBase() {
...
 }
+
+/**
+ * Validates [errorCode] and [expectedValues] for given exception [ex].
+ *
+ * @param errorCode expected errorCode
+ * @param ex actual exception thrown by test
+ * @param expectedValues expected values for errorContext
+ */
+private fun <T : SqlException> SoftAssertions.checkErrorAndErrorContext(
```
Comment (dlurton) [inline lang/test/org/partiql/lang/syntax/SqlParserTestBase.kt:238]: Same with these functions.

**Review by alancai98 — COMMENTED** [review]
I like the direction this refactor took and the efforts you made and documenting the technical debt. Left mostly minor comments and some questions about some deleted tests.

File: lang/src/org/partiql/lang/eval/NodeMetadata.kt [hunk 8]
```diff
@@ -1,63 +0,0 @@
...
-/*
- * WARNING: This whole file is intended as a non intrusive way to preserve the meta nodes information during
- * evaluation so we can include line number and column number in EvaluationExceptions. This is not a replacement for
- * properly populating the error context
- */
-
-/**
- * Holds the expression line number and column number.
- */
-data class NodeMetadata(val line: Long, val column: Long) {
```
[the remark this answers is not in the document]
Comment (lziq) [reply lang/src/org/partiql/lang/eval/NodeMetadata.kt:31]: Maybe just meta data for `ExprNode`?

File: lang/test/org/partiql/lang/eval/test/ExpectedResultFormat.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Determines the format and equivalence that is used to perform assertions on the expected result of
+ * an [EvaluatorTestCase].
+ *
+ * ## TODO
+ *
+ * Ideally, all the test cases actually use a variation of Ion equivalence that recognizes the proper semantics
+ * regarding `$partiql_bag::[]` (the order of items in bags is not relevant).  We can't have this yet because this
+ * would involve changing the expected values of up to several 10s of thousands of unit tests (most of which are
+ * generated).  None of the 4 notions of equality utilized by the integration tests today are quite ideal.
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/ExpectedResultFormat.kt:12]: I think you'd really want to keep *none* of them, if I'm honest.  I think the ideal equivalence to use for the expected values is Ion's, but with with support for bag semantics.

File: lang/test/org/partiql/lang/TestBase.kt (hunk 2 above)
Comment (dlurton) [reply lang/test/org/partiql/lang/TestBase.kt:26]: See #576 and #577.

File: lang/test/org/partiql/lang/eval/test/EvaluatorTestAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * - Tests the query planner and physical plan compiler instead.
+ * - [EvaluatorTestAdapter] that delegates multiple other implementations, allowing multiple implementations to be
+ * treated as one.
+ */
+interface EvaluatorTestAdapter {
+    /**
+     * Runs an [EvaluatorTestCase].  This is intended to be used by parameterized tests.
+     *
+     * Also runs additional assertions performed by [commonAssertions].
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/EvaluatorTestAdapter.kt:24]: Something I later moved into the other implementations `EvaluatorTestAdapter` (kdoc is out of date... fixing)

**Comment by am357** [comment]
> The purpose of this PR is to prepare the unit test suite to be able to test multiple compilers.
> 
> I have confidence in this changes because the code coverage report differs by only a fraction of a percent with these changes.
> 
> You'll want to check this out locally and start with the `EvaluatorTestBase` (ETB) class which was entirely rewritten, then look at `EvaluatorTestAdapter` and implementations of it. The other changes are (more or less) mechanical refactors that call into the new ETB functions (the bulk of the diffs in this PR). **However, I avoided changing the test cases themselves like the plague.**
> 
> For ETB, there used to be about 8 protected functions (including overloads) that were for:
> 
>     * "happy path" scenarios (i.e. given a query and a session, assert a particular value was returned), or...
> 
>     * error scenarios (i.e. given the same, assert a particular error was thrown)
> 
> 
> Each of these functions were slightly different. I replaced all of these functions in ETB with `runEvalutorTestCase` and `runEvaluatorErrorTestCase` which have all the capabilities of the prior functions and updated all of the old call sites to call one of these two functions. One of the challenges here is that, for the "happy path" tests, there are 4 different formats for expected values (see: `ExpectedResultFormat` to see how I handled this... as mentioned above, I avoided changing the test cases themselves... otherwise, this would have been far too big of a task for me to handle alone).
> 
> Each of the new ETB functions creates an instance of `EvaluatorTestCase` or `EvaluatorErrorTestCase` which contain the data for the test case and passes it to the appropriate function on `EvaluatorTestAdapter` to perform the actual tests and assertions.
> 
> `EvaluatorTestAdapter` is an interface with a few implementations--one runs the test in multiple other adapters, one that runs the test using the AST evaluator, one that tests the legacy serializers, etc... In the near future, an implementation of `EvaluatorTestAdapter` will test the new query planner & physical plan compiler.
> 
> Also, for the test DSL ([example use](https://github.com/partiql/partiql-lang-kotlin/blob/51e7da7b5e63e45f01c4df101168b2117a17a2d1/lang/test/org/partiql/lang/eval/EvaluatorTestSuite.kt#L106)) was also are modified to use a `EvaluatorTestAdapter`.
> ## One Idea for Future Use of `EvaluatorTestAdapter`
> 
> In theory, an implementation of `EvaluatorTestAdapter` could be created that emits test case data as Ion, possibly as input to a "compliance suite" of some sort. This would be a first step toward converting these unit tests to "test scripts" that could then be used to test other implementations of PartiQL.
> 
> By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Thank you for the detailed explanation; it helps a lot in reviewing this PR.

File: lang/test/org/partiql/lang/eval/test/EvaluatorTestAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    fun runEvaluatorTestCase(tc: EvaluatorTestCase, session: EvaluationSession)
+
+    /**
+     * Runs an [EvaluatorErrorTestCase].  This should be the normal entry point for parameterized error test cases.
+     *
+     * If the [EvaluatorErrorTestCase.expectedErrorCode] has [ErrorBehaviorInPermissiveMode.RETURN_MISSING] set,
+     * evaluates and asserts no [SqlException] was thrown and the return value is equal to
+     * [EvaluatorErrorTestCase.expectedPermissiveModeResult] (PartiQL equivalence.)  Otherwise, the error assertions
+     * are the same as [TypingMode.LEGACY]
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/EvaluatorTestAdapter.kt:36]: This is gone anyhow since the entire kdoc was out of date.

File: lang/test/org/partiql/lang/eval/test/ExpectedResultFormat.kt [hunk 12]
```diff
@@ -0,0 +1,60 @@
+package org.partiql.lang.eval.test
+
+/**
+ * Determines the format and equivalence that is used to perform assertions on the expected result of
+ * an [EvaluatorTestCase].
+ *
+ * ## TODO
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/ExpectedResultFormat.kt:7]: The "TODO" applies to the multiple paragraphs that follow.  Will try to find a way to make this more clear.

[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/ExpectedResultFormat.kt:7]: I updated #560 with this text, after improving it a bit, added a link to it here and removed this lengthy comment.

File: lang/test/org/partiql/lang/util/SoftAssertionsExtensions.kt [hunk 13]
```diff
@@ -0,0 +1,90 @@
+package org.partiql.lang.util
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/util/SoftAssertionsExtensions.kt:1]: Yes, thanks.

File: lang/test/org/partiql/lang/eval/test/EvaluatorTestCase.kt [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * The query to be evaluated.
+     */
+    override val query: String,
+
+    /**
+     * AN expression which will be evaluated to determine the set of expected values that should match the result
+     * of [query].
+     */
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/test/EvaluatorTestCase.kt:27]: done.

File: lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt [hunk 15]
```diff
@@ -4,56 +4,47 @@ import org.junit.Test
...
-            ExprFunctionTestCase("make_time(23, null, 59.999999999)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, null)", "NULL"),
-            ExprFunctionTestCase("make_time(null, 59, 59.999999999, 1080)", "NULL"),
-            ExprFunctionTestCase("make_time(23, null, 59.999999999, 1080)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, null, 1080)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, 59.999999999, null)", "NULL"),
-            ExprFunctionTestCase("make_time(missing, 59,59.999999999, 1080)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, missing, 1080)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, 59.999999999, missing)", "NULL"),
-            ExprFunctionTestCase("make_time(23, 59, missing, null)", "NULL")
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt:43]: Yes--they are redundant because null propagation is being implemented at the compiler level, not the function, and therefore these tests weren't even hitting the `make_time` `ExprFunction` implementation.  That functionality is being [tested elsewhere](https://github.com/partiql/partiql-lang-kotlin/blob/18c06a89b739bcb497795fe66cf6739c6efc270b/lang/test/org/partiql/lang/eval/EvaluatingCompilerUnknownValuesTest.kt#L29-L98).

File: lang/test/org/partiql/lang/eval/builtins/functions/MakeDateEvaluationTest.kt [hunk 16]
```diff
@@ -4,56 +4,47 @@ import org.junit.Test
...
-            ExprFunctionTestCase("make_date(MISSING, 02, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(MISSING, 02, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(2021, MISSING, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(2021, 02, MISSING)", "NULL"),
-            ExprFunctionTestCase("make_date(NULL, MISSING, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(MISSING, NULL, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(MISSING, 02, NULL)", "NULL"),
-            ExprFunctionTestCase("make_date(NULL, NULL, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(NULL, NULL, 28)", "NULL"),
-            ExprFunctionTestCase("make_date(MISSING, MISSING, MISSING)", "NULL"),
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/builtins/functions/MakeDateEvaluationTest.kt:42]: Same.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt [hunk 17]
```diff
@@ -112,15 +109,10 @@ class EvaluatingCompilerDateTimeTests : EvaluatorTestBase() {
             case("TIME (5) WITH TIME ZONE '12:24:12.123678+05:30'", "12:24:12.12368+05:30", TimeForValidation(12, 24, 12, 123680000, 5, 330)),
             case("TIME (2) WITH TIME ZONE '12:59:59.135-05:30'", "12:59:59.14-05:30", TimeForValidation(12, 59, 59, 140000000, 2, -330)),
             case("TIME (2) WITH TIME ZONE '12:59:59.134-05:30'", "12:59:59.13-05:30", TimeForValidation(12, 59, 59, 130000000, 2, -330)),
-            case("TIME '12:25:12.123456' IS TIME", "true"),
-            case("TIME (2) '01:01:12' IS TIME", "true"),
-            case("TIME WITH TIME ZONE '12:25:12.123456' IS TIME", "true"),
-            case("TIME (2) WITH TIME ZONE '01:01:12' IS TIME", "true"),
-            case("'01:01:12' IS TIME", "false"),
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt:119]: IMO, this is the wrong place for these tests.  The correct place is [EvaluatingCompilerIsTests](https://github.com/partiql/partiql-lang-kotlin/blob/18c06a89b739bcb497795fe66cf6739c6efc270b/lang/test/org/partiql/lang/eval/EvaluatingCompilerIsTests.kt), however, there are no equivalent tests there.  When I next push, either these tests will be restored or I'll add tests to the correct location.

[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt:119]: Nice catch btw. Thanks.

File: lang/test/org/partiql/lang/eval/CastTestBase.kt [hunk 18]
```diff
@@ -1323,11 +1313,6 @@ abstract class CastTestBase : EvaluatorTestBase() {
             ),
             // Error cases for TIME
             listOf(
-                case("TIME '23:12:12.1267'", "TIME (-1)", ErrorCode.PARSE_INVALID_PRECISION_FOR_TIME),
-                case("TIME '23:12:12.1267'", "TIME (1, 2)", ErrorCode.PARSE_CAST_ARITY),
-                case("TIME '23:12:12.1267'", "TIME (1, 2) WITH TIME ZONE", ErrorCode.PARSE_CAST_ARITY),
-                case("TIME '23:12:12.1267-05:30'", "TIME (10) WITH TIME ZONE", ErrorCode.PARSE_INVALID_PRECISION_FOR_TIME),
-                case("TIME '23:12:12.1267+05:30'", "TIME (10)", ErrorCode.PARSE_INVALID_PRECISION_FOR_TIME),
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/CastTestBase.kt:1330]: Yes - why on earth did we have parser tests (look at the error code) in the cast tests? Also, they were redundant with [these](https://github.com/partiql/partiql-lang-kotlin/blob/3febaeec46a0b703781494699c445f2034e9c6c4/lang/test/org/partiql/lang/syntax/SqlParserDateTimeTests.kt#L438-L509) tests.

File: lang/test/org/partiql/lang/eval/EvaluatorTestBase.kt [hunk 19]
```diff
@@ -21,27 +21,22 @@ import com.amazon.ion.IonType
...
+import org.partiql.lang.eval.test.LegacySerializerTestAdapter
+import org.partiql.lang.eval.test.MultipleTestAdapter
+import org.partiql.lang.eval.test.PartiqlAstExprNodeRoundTripAdapter
 import org.partiql.lang.util.asSequence
 import org.partiql.lang.util.newFromIonText
-import org.partiql.lang.util.softAssert
-import kotlin.reflect.KClass
-import kotlin.test.assertEquals
 
 /**
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/EvaluatorTestBase.kt:41]: Yes, done.  Thanks.

File: lang/src/org/partiql/lang/ast/AstDeserialization.kt [hunk 20]
```diff
@@ -12,6 +12,9 @@
  *  language governing permissions and limitations under the License.
  */
 
+// Don't need warnings about ExprNode deprecation.
+@file:Suppress("DEPRECATION")
+
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/ast/AstDeserialization.kt:17]: done.

File: lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt [hunk 21]
```diff
@@ -4,56 +4,47 @@ import org.junit.Test
...
+            ExprFunctionTestCase("make_time(0, 0, 0., 0)", "\$partiql_time::{hour:0,minute:0,second:0.,timezone_hour:0,timezone_minute:0}"),
+            ExprFunctionTestCase("make_time(23, 12, 59.12345)", "\$partiql_time::{hour:23,minute:12,second:59.12345,timezone_hour:null.int,timezone_minute:null.int}"),
+            ExprFunctionTestCase("make_time(23, 12, 59.12345, 800)", "\$partiql_time::{hour:23,minute:12,second:59.12345,timezone_hour:13,timezone_minute:20}"),
+            ExprFunctionTestCase("make_time(23, 59, 59.999999999)", "\$partiql_time::{hour:23,minute:59,second:59.999999999,timezone_hour:null.int,timezone_minute:null.int}"),
+            ExprFunctionTestCase("make_time(23, 12, 59.12345, -800)", "\$partiql_time::{hour:23,minute:12,second:59.12345,timezone_hour:-13,timezone_minute:-20}"),
+            ExprFunctionTestCase("make_time(23, 59, 59.999999999, -1080)", "\$partiql_time::{hour:23,minute:59,second:59.999999999,timezone_hour:-18,timezone_minute:0}"),
+            ExprFunctionTestCase("make_time(23, 59, 59.999999999, 1080)", "\$partiql_time::{hour:23,minute:59,second:59.999999999,timezone_hour:18,timezone_minute:0}"),
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt:37]: Actually, I tried to revert this change--but missed it.  I don't actually want to be changing the expected values of any test, if I can avoid it.

In any case, the string representation here is wrong (see #569).  Nevertheless, the expected values should be changed in a PR that fixes that issue, not this PR.

File: lang/test/org/partiql/lang/util/ExprNodeExtensions.kt [hunk 22]
```diff
@@ -0,0 +1,14 @@
+// Don't need warnings about ExprNode deprecation.
+@file:Suppress("DEPRECATION")
+
+package org.partiql.lang.util
+
+import org.partiql.lang.ast.ExprNode
+import org.partiql.lang.ast.passes.MetaStrippingRewriter
+
+@Suppress("DEPRECATION")
```
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/test/org/partiql/lang/util/ExprNodeExtensions.kt:9]: Removed.

File: lang/src/org/partiql/lang/eval/NodeMetadata.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (dlurton) [reply lang/src/org/partiql/lang/eval/NodeMetadata.kt:31]: `NodeMetadata` is a holdover from the old V0 days and wasn't used by an production code, or any other part of the public API at all, FYI.  I'll be sure to include that in the squashed commit message.

> Also, the package name "test" in "lang/test/org/partiql/lang/eval/test" seems a bit generic. Consider changing it to "test-core" or something?

I will use "evaluatortestframework" which seems long but actually compiles with the [Google java stile guide](https://google.github.io/styleguide/javaguide.html#s5.2.1-package-names) (FWIW.)

File: lang/test/org/partiql/lang/TestBase.kt (hunk 2 above)
Comment (am357) [reply lang/test/org/partiql/lang/TestBase.kt:26]: > See #576 and #577.
Would be great to add the refs to the code.

File: lang/test/org/partiql/lang/eval/builtins/functions/MakeDateEvaluationTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply lang/test/org/partiql/lang/eval/builtins/functions/MakeDateEvaluationTest.kt:42]: Similar comment to the `make_time`. I think it would be good to assert somewhere that `make_date` is a null-propagating function.

File: lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (alancai98) [reply lang/test/org/partiql/lang/eval/builtins/functions/MakeTimeEvaluationTest.kt:43]: Even though null propagation is tested elsewhere, I think having some test ensuring that `make_time` is a null-propagating function could be valuable. Also there was a test on L32 that was deleted not related to null propagation.

**Review by alancai98 — COMMENTED** [review]
Left a couple followup comments. Otherwise looks good to me since last review

**Review by alancai98 — APPROVED** [review]
